src/Microsoft.Health.Fhir.SqlServer/Features/ChangeFeed/SqlServerFhirResourceChangeDataStore.cs

```diff
@@ -0,0 +1,159 @@
+
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Data.SqlClient;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Abstractions.Data;
+using Microsoft.Health.Fhir.Core.Models;
+using Microsoft.Health.SqlServer;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.ChangeFeed
+{
+    /// <summary>
+    /// A data store that provides functionalities to get resource changes and resource types for change feed sources.
+    /// </summary>
+    public class SqlServerFhirResourceChangeDataStore : IChangeFeedSource<IResourceChangeData>
+    {
+        private readonly ISqlConnectionFactory _sqlConnectionFactory;
+        private readonly ILogger<SqlServerFhirResourceChangeDataStore> _logger;
+
+        // dbnetlib error value for timeout expired
+        private const short TIMEOUTEXPIRED = -2;
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="SqlServerFhirResourceChangeDataStore"/> class.
+        /// </summary>
+        /// <param name="sqlConnectionFactory">The SQL Connection factory.</param>
+        /// <param name="logger">The logger.</param>
+        public SqlServerFhirResourceChangeDataStore(ISqlConnectionFactory sqlConnectionFactory, ILogger<SqlServerFhirResourceChangeDataStore> logger)
+        {
+            EnsureArg.IsNotNull(sqlConnectionFactory, nameof(sqlConnectionFactory));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _sqlConnectionFactory = sqlConnectionFactory;
+            _logger = logger;
+        }
+
+        /// <summary>
+        ///  Returns the number of resource change records from startId.
+        /// </summary>
+        /// <param name="startId">The start id of resource change records to fetch. The start id is inclusive.</param>
+        /// <param name="pageSize">The page size for fetching resource change records.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>Resource change data rows.</returns>
+        /// <exception cref="System.ArgumentOutOfRangeException">Thrown if startId or pageSize is less than zero.</exception>
+        /// <exception cref="System.InvalidOperationException">Thrown when a method call is invalid for the object's current state.</exception>
+        /// <exception cref="Microsoft.Data.SqlClient.SqlException">Thrown when SQL Server returns a warning or error.</exception>
+        /// <exception cref="System.TimeoutException">Thrown when the time allotted for a process or operation has expired.</exception>
+        /// <exception cref="System.Exception">Thrown when errors occur during execution.</exception>
+        public async Task<IReadOnlyCollection<IResourceChangeData>> GetRecordsAsync(long startId, int pageSize, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsGte(startId, 0, nameof(startId));
```

Comment:
>startId [](http://example.com/codeflow?start=28&length=7)

supernit: 
`Id bigint IDENTITY(1,1)`
it starts from 1, so having start id equal to 0 is same as letting it be negative (doesn't break anything, but doesn't do anything as well.)

Reply:
Nice catch. Changed the start id validation as greater than or equal to 1.